lucet-runtime/lucet-runtime-tests/src/timeout.rs

```diff
@@ -97,61 +110,90 @@ macro_rules! timeout_tests {
             match inst.run("infinite_loop", &[]) {
                 Err(Error::RuntimeTerminated(TerminationDetails::Remote)) => {
                     // this is what we want to see
                 }
                 res => panic!("unexpected result: {:?}", res),
             }
+            t.join().unwrap();
 
-            // after a timeout, can reset and run a normal function
-            inst.reset().expect("instance resets");
+            // Another attempt to terminate the instance will fail.
+            assert_eq!(
+                inst.kill_switch().terminate(),
+                Err(KillError::NotTerminable)
+            );
 
+            // Check that we can reset the instance and run a normal function.
+            inst.reset().expect("instance resets");
             run_onetwothree(&mut inst);
         }
 
         #[test]
-        fn timeout_before_guest() {
+        fn timeout_before_guest_runs() {
             let module = mock_timeout_module();
             let region = TestRegion::create(1, &Limits::default()).expect("region can be created");
             let mut inst = region
                 .new_instance(module)
                 .expect("instance can be created");
-
             let kill_switch = inst.kill_switch();
-            assert_eq!(kill_switch.terminate(), Err(KillError::NotTerminable));
 
-            // not being terminable, the instance still runs and is unaffected
+            // If terminated before running, the guest will be cancelled.
+            assert_eq!(kill_switch.terminate(), Ok(KillSuccess::Cancelled));
+
+            // Another attempt to terminate the instance will fail.
+            assert_eq!(
+                inst.kill_switch().terminate(),
+                Err(KillError::NotTerminable)
+            );
+
+            match inst.run("onetwothree", &[]) {
+                Err(Error::RuntimeTerminated(TerminationDetails::Remote)) => {}
+                res => panic!("unexpected result: {:?}", res),
+            }
+
+            // Check that we can reset the instance and run a normal function.
+            inst.reset().expect("instance resets");
             run_onetwothree(&mut inst);
         }
 
+        /// FIXME: This needs a subtle change.
         #[test]
-        fn timeout_after_guest() {
+        fn timeout_after_guest_runs() {
             let module = mock_timeout_module();
             let region = TestRegion::create(1, &Limits::default()).expect("region can be created");
             let mut inst = region
                 .new_instance(module)
                 .expect("instance can be created");
+            let kill_switch = inst.kill_switch();
 
+            // The killswitch will fail if the instance has already finished running.
             match inst.run("do_nothing", &[]) {
                 Ok(_) => {}
                 res => panic!("unexpected result: {:?}", res),
             }
+            assert_eq!(kill_switch.terminate(), Ok(KillSuccess::Cancelled)); // FIXME -> Invalid
```

Comment:
```suggestion
            assert_eq!(kill_switch.terminate(), Ok(KillSuccess::Cancelled));
```
`Cancelled` seems correct here, since you could try to run another function on the same instance, but if it timed out in the mean time the instance should refuse to.

Reply:
Yep, you're exactly right :slightly_smiling_face: